x-pack/plugins/code/server/security.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { Lifecycle, Request, ResponseToolkit, Server, ServerRoute } from 'hapi';
+
+export interface SecuredRoute extends ServerRoute {
+  requireRoles?: string[];
+  requireAdmin?: boolean;
+}
+
+export const ADMIN_ROLE = 'code_admin';
+
+declare module 'hapi' {
+  interface Server {
+    securedRoute(route: SecuredRoute): void;
+  }
+}
+
+export class SecureRoute {
+  constructor(readonly server: Server) {}
+
+  public install() {
+    const self = this;
+    function securedRoute(route: SecuredRoute) {
+      if (route.handler) {
+        const originHandler = route.handler as Lifecycle.Method;
+        route.handler = async (req: Request, h: ResponseToolkit, err?: Error) => {
+          if (self.isSecurityEnabledInEs()) {
+            let requiredRoles = route.requireRoles || [];
```

Comment:
I agree that we should at least be removing the hard-coded roles, and relying either only on direct index privileges to `.code-*` or on the application privileges which we'd grant to code. 

While application privileges would be preferential, they're also likely the most effort at this point as our role management screens and APIs only allow Kibana privileges to be granted, and those aren't easily extensible for application like code at this point in time.

Reply:
@legrego @kobelb we do this specific for strong security reason:

our plugin does not only just read/write elasticsearch index, one use case is the `Index API`, the API start language server for a project, send initialization request (which might be expensive) and then query some other API and use the data returned by the API to write to elasticsearch index.

if we don't do role control here, it does work, because the request to elasticsearch would fail, however that means user could freely send request to let server to do  many other works which seems to be a severe DDOS vulnerability.

Reply:
> if we don't do role control here, it does work, because the request to elasticsearch would fail, however that means user could freely send request to let server to do many other works which seems to be a severe DDOS vulnerability.

Are there any writes to the `.code-*` indices that you can do before the expensive server operations are performed? I understand that most of the writes happen after (because you need the repo to be cloned/analyzed before you can write about it), but if there is any job metadata that can be written beforehand, that would give you the gate you need to prevent unauthorized users from performing these expensive operations.

> Also, do you think this is a blocker to the initial code merge? Would it be better to do that later when there is a horizontal initiative to migrate similar usage case to purely use elasticsearch's index permission.

I can be ok with this for the short-term, as long as there is a path forward for us to take advantage of Feature Controls once they ship without a ton of effort on our end-users. Feature Controls will give you the tools you need to secure your API endpoints without hard-coded role checks. Based on what I've seen, nothing here will prevent us from using Feature Controls.

Before we release, please make sure we have adequate end-user documentation around these roles, so that users know how to configure their users/roles for access to the Code application. No need to block the merge on docs!

Additionally, please make sure that your UI gracefully handles scenarios where users do not have adequate access. I didn't test this, so I apologize if it's already there 😄

Reply:
we don't have those writes, I'll create a separate issue track the issue
https://github.com/elastic/code/issues/837